### PR TITLE
Common code support for AES-XTS

### PR DESCRIPTION
**Pull request by ifranzki**
This PR took the code from https://github.com/opencryptoki/opencryptoki/pull/606 from @vkanjur and reworked it to support multi-part operations as well.

@vkanjur I left you as the Author of the very first commit, since you originally developed it. I also kept your sign-off but added mine as well. I hope this is OK with you? 

It also adds a generic AES-XTS implementation that can be used by tokens to implement AES-XTS support. The generic implementation only requires the tokens to provide two callback functions. One to calculate the initial IV from the tweak value, and one for performing the AES-XTS full-block encryption/decryption.

Furthermore, this PR adds support for AES-XTS for the Soft and ICA tokens. Since the OpenSSL AES-XTS implementation only supports single-part operations, the generic AES implementation is used to implement multi part AES-XTS with the help of OpenSSL. For ICA, an enhanced libica providing a multi-part capable AES-XTS implementation is required. 

We will need more test vectors for AES-XTS, especially for non-full-block cases. Those should be added at a later time.

### CODE REVIEW

File: usr/lib/ica_s390_stdll/tok_struct.h
Comment (jschmidb) [inline usr/lib/ica_s390_stdll/tok_struct.h:112]: Is this intended: lines 111 and 112 contain the same function ptr?

File: usr/lib/common/mech_openssl.c
Comment (jschmidb) [inline usr/lib/common/mech_openssl.c:4250]: Indentation?

File: usr/lib/soft_stdll/soft_specific.c
Comment (jschmidb) [inline usr/lib/soft_stdll/soft_specific.c:542]: I guess the keysize was already checked before calling this routine? (only 128 and 256)

File: usr/lib/common/h_extern.h
Comment (jschmidb) [inline usr/lib/common/h_extern.h:1485]: Similar to the previous pr: I guess at least in_data is 'const' here.

File: usr/lib/common/h_extern.h
Comment (jschmidb) [inline usr/lib/common/h_extern.h:1342]: const CK_BYTE *in_data   ?

File: usr/lib/ica_s390_stdll/tok_struct.h
Comment (ifranzki) [reply usr/lib/ica_s390_stdll/tok_struct.h:112]: Oh, no, not intended :-) Although it technically does not make a difference, since it just generates a random key in the length as specified by the caller. Will fix anyway .

File: usr/lib/common/mech_openssl.c
Comment (ifranzki) [reply usr/lib/common/mech_openssl.c:4250]: fixed.

File: usr/lib/soft_stdll/soft_specific.c
Comment (ifranzki) [reply usr/lib/soft_stdll/soft_specific.c:542]: yes, that has already been checked.

File: usr/lib/common/mech_aes.c
Comment (jschmidb) [inline usr/lib/common/mech_aes.c:1287]: For shorter path length we could also check here if (in_data == NULL || in_data_len == 0) and just return

File: usr/lib/common/h_extern.h
Comment (ifranzki) [reply usr/lib/common/h_extern.h:1342]: Hm yes, should be const indeed. The problem is that the caller (encr_mgr_encrypt_update) does not define in_data as const (although it should be const there, too). 
I leave it as is for this PR, but I put it on my TODP list to update all places where arguments could be const.

File: usr/lib/common/mech_aes.c
Comment (ifranzki) [reply usr/lib/common/mech_aes.c:1287]: this has already been checked by the caller (i.e. in SC_EncryptUpdate)

Comment (ifranzki) [reply usr/lib/common/mech_aes.c:1287]: ... and `in_data_len == 0` is actually allowed. Also `in_data` may be `NULL` if `in_data_len` is zero. 
Only if `in_data_len` is not zero, then `in_data` must also be non-NULL.

File: usr/lib/common/mech_aes.c
Comment (jschmidb) [inline usr/lib/common/mech_aes.c:4218]: What was the reason to add this generic implementation? I mean, yes, it's good to have one, but was it not possible to use openssl with a bit of glue code for this?

Comment (ifranzki) [reply usr/lib/common/mech_aes.c:4218]: OpenSSL does not support multi-part operations with AES-XTS ! 

The OpenSSL man page states:

_The XTS implementation in OpenSSL does not support streaming. That is there must only be one EVP_EncryptUpdate call per EVP_EncryptInit_ex call (and similarly with the "Decrypt" functions)._

So with the generic implementation we can implement multi-part operations for the SoftToken by using OpenSSL's AES-ECB functions, also see commit message of Soft Token commit: https://github.com/opencryptoki/opencryptoki/pull/614/commits/a0a502ca6ec757cea9a7248219d086a956eb1ca9

Comment (ifranzki) [reply usr/lib/common/mech_aes.c:4218]: ... and it also helps to implement AES-XTS with multi part support for the EP11 token (with protected keys) for example. The 2 callback functions corelate to what the PCC and KM instructions do....

**Comment by vkanjur** [comment]
Approved

**Comment by ifranzki** [comment]
@jschmidb Are you OK with this PR? If so, please approve.

@vkanjur Thanks for your approval, can you please also press the Approve button? (Navigate to one commit of this PR, then the green button "Review changes", then select Approve and press submit review - for whatever reason the approve function can only be found when in a commit .....)

Once this one is merged, I will close/cancel https://github.com/opencryptoki/opencryptoki/pull/606 .

**Review by vkanjur — APPROVED** [review]

**Review by jschmidb — APPROVED** [review]
lgtm

**Pull request merged**
